Add webhook validations secret object

Anandkumar26 opened the pull request:
- Update of secrets are allowed only when the secret key
  (which is reffered by CPA) is not changed.
- Delete of secrets are allowed only when it is not being
  referred by CPA.
- Add unit tests for validating webhook for Secret
- Coverage at 95.1%

Signed-off-by: Anand Kumar <kumaranand@vmware.com>

Anandkumar26 commented:
Yet to add the kubebuilder verbs to auto-generate the RBAC for webhook.

reachjainrahul commented on config/nephe.yml:
```diff
@@ -37,18 +37,13 @@ spec:
   - name: v1alpha1
     schema:
       openAPIV3Schema:
-        description: CloudEntitySelector is the Schema for the cloudentityselectors
```
update your nephe builder image. it should now use kustomizer 4.4.1

Anandkumar26 replied:
Done

reachjainrahul commented on cmd/nephe-controller/main.go:
```diff
@@ -47,6 +56,124 @@ func init() {
+	decoder *admission.Decoder
+}
+
+// allowSecretUpdate returns nil only when secret name/namespace/key are same.
```
create a separate file for it.. Dont add in main

Anandkumar26 replied:
Done.

reachjainrahul commented on cmd/nephe-controller/main.go:
```diff
@@ -47,6 +56,124 @@ func init() {
+
+// allowSecretUpdate returns nil only when secret name/namespace/key are same.
+func (v *secretValidator) allowSecretUpdate(new *corev1.Secret, old *corev1.Secret, key string) error {
+	setupLog.Info("Checking secrets are changed")
```
shouldnt be using setupLog

Anandkumar26 replied:
Added a new logger instance for secret webhook.

reachjainrahul commented on cmd/nephe-controller/main.go:
```diff
@@ -47,6 +56,124 @@ func init() {
+	case admissionv1.Create:
+		return admission.Allowed("")
+	case admissionv1.Update:
+		newSecret := &corev1.Secret{}
```
Create separate routine for update and delete case..

Anandkumar26 replied:
Done

reachjainrahul commented on cmd/nephe-controller/main.go:
```diff
@@ -47,6 +56,124 @@ func init() {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		err, cpa := v.fetchCPABySecret(newSecret, oldSecret, false)
+		if err != nil && strings.Contains(err.Error(), "failed to get cpa list") {
```
what about other errors?

Anandkumar26 replied:
This is to handle a case where we were unable to get CPA list (for whatever reason).
For other scenarios, we will always return a CPA CR or error will be nil.

reachjainrahul commented on cmd/nephe-controller/main.go:
```diff
@@ -47,6 +56,124 @@ func init() {
+			setupLog.Error(err, "Error while decoding secretValidator", "secretValidator", req.Name)
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		err, cpa := v.fetchCPABySecret(newSecret, oldSecret, false)
```
IMO, this should be called, only if credentials in secret is changed.. If label is updated, we should allow it.

Anandkumar26 replied:
The CPA CR uses a secret key reference. So unless we parse the secret data map and compare the data map keys with CPA secret key, we won't know if this is something we are interested in.

So the check has to be inside fetchCPABySecret.

If any other field of Secret CR is updated, it is allowed.

reachjainrahul commented on cmd/nephe-controller/main.go:
```diff
@@ -47,6 +56,124 @@ func init() {
+			return admission.Denied(fmt.Sprintf("%s used by cpa %s", err.Error(), cpa.Name))
+		}
+	default:
+		setupLog.Info("Invalid admission webhook request")
```
you are allowing invalid admission request

reachjainrahul commented on cmd/nephe-controller/main.go:
```diff
@@ -47,6 +56,124 @@ func init() {
+	options := &controllerclient.ListOptions{Limit: 1000}
+	err := v.client.List(context.TODO(), cpaList, options)
+	if err != nil {
+		return fmt.Errorf("failed to get cpa list, err:%v", err), nil
```
dont use CPA, use full name..

Anandkumar26 replied:
Done

reachjainrahul commented on cmd/nephe-controller/main.go:
```diff
@@ -47,6 +56,124 @@ func init() {
+// allowSecretUpdate returns nil only when secret name/namespace/key are same.
+func (v *secretValidator) allowSecretUpdate(new *corev1.Secret, old *corev1.Secret, key string) error {
+	setupLog.Info("Checking secrets are changed")
+	if new.Name != old.Name {
```
is this even a validate case?

Anandkumar26 replied:
Yes, its not valid. Name and Namespace changes are not allowed for a CR

reachjainrahul commented on cmd/nephe-controller/main.go:
```diff
@@ -18,17 +18,18 @@ import (
 	controllers "antrea.io/nephe/pkg/controllers/cloud"
 	"antrea.io/nephe/pkg/logging"
+
+	"k8s.io/apimachinery/pkg/runtime"
```
system, third party and then antrea imports... move this up

Anandkumar26 replied:
Done

reachjainrahul commented on cmd/nephe-controller/secret_webhook.go:
```diff
@@ -0,0 +1,169 @@
+	crdv1alpha1 "antrea.io/nephe/apis/crd/v1alpha1"
+)
+
+// SecretValidator is used to validate Secret
```
end with .

Anandkumar26 replied:
Done

reachjainrahul commented on cmd/nephe-controller/secret_webhook.go:
```diff
@@ -0,0 +1,169 @@
+
+// getCPABySecret returns nil only when the Secret is not used by
+// CloudProvideAccount,
+// Otherwise, an appropriate error message and the corresponding
```
nit, club with previous line

Anandkumar26 replied:
Done

reachjainrahul commented on cmd/nephe-controller/secret_webhook.go:
```diff
@@ -0,0 +1,169 @@
+	}
+
+	for _, cpa := range cpaList.Items {
+		v.Log.Info("Checking CloudProvideAccount", "Name", cpa.Name, "Namespace", cpa.Namespace)
```
move this to debug log

Anandkumar26 replied:
Done

reachjainrahul commented on cmd/nephe-controller/secret_webhook.go:
```diff
@@ -0,0 +1,169 @@
+			if cpa.Spec.AWSConfig.SecretRef.Name == oldSecret.Name && cpa.Spec.AWSConfig.SecretRef.Namespace == oldSecret.Namespace {
+				if isDelete {
+					// Delete is not allowed.
+					v.Log.Info("Cannot delete", "Secret", oldSecret.Name, "CloudProvideAccount", cpa.Name)
```
update this logging, else error message and this log is the same.. I believe you want to highlight its used by CPA

Anandkumar26 replied:
I just want to have a log message which captures an attempt was made to delete/update secret which is used by CPA.
Sure, I will change it to ERROR message.

Anandkumar26 replied:
Changing the logs type to ERROR, will dump the call-stack. Also this is not an error, so IMO it should be an INFO message only.
 
2022-09-07T08:53:46.430Z        ERROR   webhook.Secret  webhook/secret_webhook.go:56    Secret resource cannot be deleted       {"Secret": "aws-account-creds", "CloudProviderAccount": "cloudprovideraccount-aws-sample"}
antrea.io/nephe/pkg/apiserver/webhook.(*SecretValidator).Handle
        /usr/src/antrea.io/nephe/pkg/apiserver/webhook/secret_webhook.go:56
sigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).Handle
        /tmp/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/webhook/admission/webhook.go:146
sigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).ServeHTTP
        /tmp/gopath/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/webhook/admission/http.go:98
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1
        /tmp/gopath/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:40
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1
        /tmp/gopath/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:117
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2
        /tmp/gopath/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:84
net/http.HandlerFunc.ServeHTTP
        /usr/local/go/src/net/http/server.go:2109
net/http.(*ServeMux).ServeHTTP
        /usr/local/go/src/net/http/server.go:2487
net/http.serverHandler.ServeHTTP
        /usr/local/go/src/net/http/server.go:2947
net/http.(*conn).serve
        /usr/local/go/src/net/http/server.go:1991

reachjainrahul commented on cmd/nephe-controller/secret_webhook.go:
```diff
@@ -0,0 +1,169 @@
+
+// allowSecretUpdate returns true only when Secret data key is unchanged
+func (v *SecretValidator) allowSecretUpdate(new *corev1.Secret, old *corev1.Secret, key string) bool {
+	if changed := string(new.Data[key]) != string(old.Data[key]); changed {
```
do you really need key to be passed? isnt just compare data field sufficient?

Anandkumar26 replied:
As discussed this is required, as Secret Data is a map and it can contain multiple key, value pairs. CPA uses only particular key from this map. So we have to match with key in CPA CR.

reachjainrahul commented on cmd/nephe-controller/secret_webhook.go:
```diff
@@ -0,0 +1,169 @@
+// CloudProvideAccount,
+// Otherwise, an appropriate error message and the corresponding
+// CloudProvideAccount CR will be returned.
+func (v *SecretValidator) getCPABySecret(newSecret *corev1.Secret, oldSecret *corev1.Secret,
```
I think you should split this into 2. this function should just return CPA, and in the caller do extra validation, if update or delete is allowed

Anandkumar26 replied:
Done

reachjainrahul commented on cmd/nephe-controller/secret_webhook.go:
```diff
@@ -0,0 +1,169 @@
+	}
+
+	err, cpa := v.getCPABySecret(newSecret, oldSecret, false)
+	if err != nil && strings.Contains(err.Error(), "failed to get CloudProviderAccount list") {
```
you are doing pattern match on error. may be define a string const and use at both places..

Anandkumar26 replied:
Done

reachjainrahul commented on config/nephe.yml:
```diff
@@ -37,18 +37,13 @@ spec:
+        description: CloudEntitySelector is the Schema for the cloudentityselectors API.
         properties:
           apiVersion:
-            description: 'APIVersion defines the versioned schema of this representation
```
Upgrade your nephe builder image.. You shoulnt be seeing this change

Anandkumar26 replied:
Done

reachjainrahul commented on Makefile:
```diff
@@ -7,8 +7,8 @@ CRD_OPTIONS ?= "crd"
 DOCKER_GOCACHE=/tmp/gocache
-GENERATE_CODE_LIST={$$(go list ./... | grep -e apis/crd -e apis/runtime | paste -s -d, -)}
-GENERATE_MANIFEST_LIST={$$(go list ./... | grep apis/crd | paste -s -d, -)}
+GENERATE_CODE_LIST={$$(go list ./... | grep -e apis/crd -e apis/runtime| paste -s -d, -)}
```
nit, give space

Anandkumar26 replied:
Done

reachjainrahul commented on Makefile:
```diff
@@ -7,8 +7,8 @@ CRD_OPTIONS ?= "crd"
-GENERATE_CODE_LIST={$$(go list ./... | grep -e apis/crd -e apis/runtime | paste -s -d, -)}
-GENERATE_MANIFEST_LIST={$$(go list ./... | grep apis/crd | paste -s -d, -)}
+GENERATE_CODE_LIST={$$(go list ./... | grep -e apis/crd -e apis/runtime| paste -s -d, -)}
+GENERATE_MANIFEST_LIST={$$(go list ./... | grep -E 'apis/crd|pkg/apiserver/webhook' | paste -s -d, -)}
```
nit here too

Anandkumar26 replied:
This is a regular expression, so space cannot be added here.

reachjainrahul commented on cmd/nephe-controller/main.go:
```diff
@@ -21,12 +21,14 @@ import (
 	crdv1alpha1 "antrea.io/nephe/apis/crd/v1alpha1"
 	runtimev1alpha1 "antrea.io/nephe/apis/runtime/v1alpha1"
 	"antrea.io/nephe/pkg/apiserver"
+	apiserverwebhook "antrea.io/nephe/pkg/apiserver/webhook"
```
rename to nephewebhook

Anandkumar26 replied:
Done

reachjainrahul commented on pkg/apiserver/webhook/nephe_controller_suite_test.go:
```diff
@@ -0,0 +1,13 @@
+package webhook_test
```
Copyright is missing

Anandkumar26 replied:
Done

reachjainrahul commented on pkg/apiserver/webhook/secret_webhook.go:
```diff
@@ -0,0 +1,167 @@
+	case admissionv1.Delete:
+		return v.validateDelete(req)
+	default:
+		v.Log.Info("Invalid admission webhook request")
```
In case of default, you aren't returning error

Anandkumar26 replied:
Should we deny this request?

reachjainrahul commented on pkg/apiserver/webhook/secret_webhook.go:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	for _, cpa := range cpaList.Items {
+		v.Log.V(1).Info("Checking CloudProvideAccount", "Name", cpa.Name,
```
Move this to debug log..

Anandkumar26 replied:
This is debug log. At level 1 - implies its debug. 

```
2022-09-07T08:28:41.400Z        DEBUG   webhook.Secret  webhook/secret_webhook.go:156   Checking CloudProvideAccount    {"Name": "cloudprovideraccount-aws-sample", "Namespace": "sample-ns"}
2022-09-07T08:28:41.400Z        INFO    webhook.Secret  webhook/secret_webhook.go:56    Secret resource cannot be deleted       {"Secret.Name": "aws-account-creds", "CloudProviderAccount.Name": "cloudprovideraccount-aws-sample"}
```

reachjainrahul commented on pkg/apiserver/webhook/secret_webhook.go:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	err, cpa := v.getCPABySecret(oldSecret)
+	if err != nil && strings.Contains(err.Error(), errorMsg00001) {
```
you will return success case, if its an error and doesnt match a pattern. Please handle it too. I know now your code returns on that error, very easy to miss it in future

Anandkumar26 replied:
sure. I get that.

reachjainrahul commented on pkg/apiserver/webhook/secret_webhook.go:
```diff
@@ -0,0 +1,167 @@
+	if cpa != nil {
+		v.Log.Info("Secret resource cannot be deleted", "Secret", oldSecret.Name,
+			"CloudProviderAccount", cpa.Name)
+		return admission.Denied(fmt.Sprintf("cannot delete Secret %s used by CloudProviderAccount %s",
```
Line 162, 163  convey the same info. Can you remove at time 162

Anandkumar26 replied:
If we skip L162, then in production case, we wont have any information about the denying the request.

Anandkumar26 replied:
Discussed offline. Decided to change the log message to ERROR level and rephrase the error.

reachjainrahul commented on pkg/apiserver/webhook/nephe_controller_test.go:
```diff
@@ -0,0 +1,940 @@
+				fmt.Printf("Failed to inject decoder, err %v\n", err)
+			}
+			response := SecretValidatorTest1.Handle(context.Background(), s1Req)
+			fmt.Printf("Got admission response %+v\n", response)
```
why use standard formatting. Doesnt Ginkgo provide logging.

Anandkumar26 replied:
It provides. v2 version of ginkgo gives same options as fmt. 
While the v1 version which we are using, only has one option., as shown below.

```
_, _ = GinkgoWriter.Write([]byte(fmt.Sprintf("Got admission response %+v\n", response)))
```

For now, I have changed log to use the above format, but in future we could upgrade ginkgo to v2.

reachjainrahul commented on pkg/apiserver/webhook/nephe_controller_test.go:
```diff
@@ -0,0 +1,940 @@
+		It("Validate Secret delete with dependent AWS CPA", func() {
+			fmt.Printf("Creating Secret [%s, %s]\n", s1.Name, s1.Namespace)
+			err = fakeClient.Create(context.Background(), s1)
+			if err != nil {
```
use EXPECT other than this check

Anandkumar26 replied:
Done

reachjainrahul commented:
/nephe-test-e2e-kind

Anandkumar26 commented:
The nephe-test-e2e-aws and nephe-test-e2e-azure has passed. 
Rebased on top of main branch.

Changed function descriptions for few functions, and updated comment in the test file.
